retrofit/src/main/java/retrofit2/RequestBuilder.java

```diff
@@ -107,19 +106,16 @@ private static String canonicalize(String input, boolean alreadyEncoded) {
     return input;
   }
 
-  private static void canonicalize(Buffer out, String input, int pos, int limit,
-      boolean alreadyEncoded) {
+  private static void canonicalizeForPath(Buffer out, String input, int pos, int limit,
+                                          boolean alreadyEncoded) {
     Buffer utf8Buffer = null; // Lazily allocated.
     int codePoint;
     for (int i = pos; i < limit; i += Character.charCount(codePoint)) {
       codePoint = input.codePointAt(i);
       if (alreadyEncoded
-          && (codePoint == '\t' || codePoint == '\n' || codePoint == '\f' || codePoint == '\r')) {
+          && shouldSkipCodePoint(codePoint)) {
         // Skip this character.
-      } else if (codePoint < 0x20
-          || codePoint >= 0x7f
-          || PATH_SEGMENT_ENCODE_SET.indexOf(codePoint) != -1
-          || (codePoint == '%' && !alreadyEncoded)) {
+      } else if (!alreadyEncoded && shouldEncodeCodePoint(codePoint)) {
```

Comment:
I agree. I think there is a disagreement between semantics, and the documentation/code not aligning with my (mis)interpretation. Are you able to see where my confusion is coming from -- why i consider re-encoding to have `alreadyEncoded = true` but also canonicalize/encode `/` to `%2F`?

If you don't want to current behavior of the encoded parameter to change, then thats fine. If you don't see why someone may misunderstand the encoded parameter thats a different thing. As a way to prevent further misunderstandings can we at least agree to update the documentation of the encoded parameter?

```
/**
 * Specifies whether the argument value to the annotated method parameter is already URL encoded.
 */
```

Thanks for the explanation of Retrofit/OkHttp using an ordered list or path segments. I suppose it would have been clearer to me if i hadn't seen the url being build with string replacements...

```
relativeUrl = relativeUrl.replace("{" + name + "}", canonicalizeForPath(value, encoded));
```
I will spend the rest of the day looking how to add in muti-segment support, instead of the changing the behavior of `encoded` if you would rather... Would you like this pull request to be closed?